libs/specialPushNotificationManager.js

```diff
@@ -0,0 +1,92 @@
+var sprintf = require('sprintf-js');
+
+var db, queue, pushNotifications, done, habitrpgUsers, notificationBuckets, timezoneQuery;
+
+var pageLimit = 30;
+
+function processUsersWithDevices(users) {
+  if (users.length === 0) {
+    done();
+    return;
+  }
+
+  var buckets = [];
+  var noMessageUsers = [];
+
+  notificationBuckets.forEach(notificationBucket => {
+    buckets.push({
+      identifier: notificationBucket.identifier,
+      title: notificationBucket.title,
+      message: notificationBucket.message,
+      users: []
+    })
+  });
+
+  users.forEach(user => {
+    var id = Math.random()*100;
+    if (id <= 5) {
+      noMessageUsers.push(user);
+    } else {
+      var bucketId = Math.floor(id / 100 * buckets.length);
+      buckets[bucketId].users.push(user);
+    }
+  });
+
+  buckets.forEach(bucket => {
+    bucket.users.forEach(user => {
+      var details = {
+        identifier: "wonChallenge", //We still need a generic notification type for android.
+        title: bucket.title,
+        message: sprintf.sprintf(bucket.message, user.profile.name)
+      };
+      pushNotifications.sendNotification(user, details);
+    });
+  });
+
+  if (users.length === pageLimit) {
+    var lastUser = users[users.length - 1];
+    sendPushnotifications(lastUser._id);
+  } else {
+    done();
+  }
+};
+
+function sendPushnotifications(lastId) {
+  var query = {
+    'pushDevices': {'$gt': []},
+    'preferences.timezoneOffset': timezoneQuery
+  };
+
+  if (lastId) {
+    query._id = {
+      $gt: lastId
+    }
+  }
+
+  habitrpgUsers.find(query, {
+    sort: {_id: 1},
+    limit: pageLimit,
+    fields: ['_id', 'pushDevices', 'profile']
+  })
+    .then(processUsersWithDevices)
+    .catch(function (err) {
```

Comment:
@vIiRuS you have to do it manually because kue doesn't support promises. The worker function gets passed `job, done` parameters. You have to call `done()` when the job has finished without any error and `done(err)` if an error occurred